### PR TITLE
enable pipeline checkpoint loading mode

### PR DESCRIPTION
**Pull request by leiwen83**
In cpu ram limited machine, loading checkpoint at the start up may cause oom as all rank in the same node are loading the opt state in the same time. So for this scenario, we make a choice that loading checkpoint could be made pipeline way.

### CODE REVIEW

**Comment by leiwen83** [comment]
@microsoft-github-policy-service agree [company="{iQiyi}"]



2023年5月30日 09:20，microsoft-github-policy-service[bot] ***@***.***> 写道：


@microsoft-github-policy-service agree [company="{your company}"]

**Comment by leiwen83** [comment]
@microsoft-github-policy-service agree



2023年5月30日 09:22，microsoft-github-policy-service[bot] ***@***.***> 写道：


@microsoft-github-policy-service agree

**Comment by leiwen83** [comment]
@tjruwase Would you mind take a look at this patch, since I see you are actively in zero developing.

Thanks~

**Comment by tjruwase** [comment]
@leiwen83, thanks, this looks awesome. Could you please add this option to the unit tests to get some test coverage?

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -2804,10 +2814,17 @@ def _load_zero_checkpoint(self, load_dir, tag, load_optimizer_states=True):
             if zero_sd_list is None:
                 return False
 
-        self.optimizer.load_state_dict(state_dict_list=zero_sd_list,
-                                       load_optimizer_states=load_optimizer_states,
-                                       load_from_fp32_weights=self.zero_load_from_fp32_weights(),
-                                       checkpoint_folder=checkpoint_folder)
+        if load_serial is not None:
```
Comment (tjruwase) [inline deepspeed/runtime/engine.py:2817]: If this `if` condition needed since we are passing `load_serial` along?

Comment (leiwen83) [reply deepspeed/runtime/engine.py:2817]: Yep, it could be simplified. I would send another patch to fix it, since I found original patch already merged?

**Comment by leiwen83** [comment]
> @leiwen83, thanks, this looks awesome. Could you please add this option to the unit tests to get some test coverage?

I see zero related unit test located at tests/unit/runtime/zero/. 
Per my understanding, maybe I need to add into tests/unit/runtime/zero/test_zero_config.py to verify this new configuration?

**Comment by tjruwase** [comment]
> > @leiwen83, thanks, this looks awesome. Could you please add this option to the unit tests to get some test coverage?
> 
> I see zero related unit test located at tests/unit/runtime/zero/. Per my understanding, maybe I need to add into tests/unit/runtime/zero/test_zero_config.py to verify this new configuration?

Yes, that is what I had in mind. Thanks!

**Comment by leiwen83** [comment]
> > > @leiwen83, thanks, this looks awesome. Could you please add this option to the unit tests to get some test coverage?
> > 
> > 
> > I see zero related unit test located at tests/unit/runtime/zero/. Per my understanding, maybe I need to add into tests/unit/runtime/zero/test_zero_config.py to verify this new configuration?
> 
> Yes, that is what I had in mind. Thanks!

I added the unit test code and modify the load_serial logic in the latest patch.

**Comment by leiwen83** [comment]
> @leiwen83, apologies for not being clear, unit tests of the pipeline option should also be added to https://github.com/microsoft/DeepSpeed/blob/master/tests/unit/checkpoint/test_zero_optimizer.py

Sorry for late, test_zero_optimizer.py unit test has been added.

**Review by tjruwase — APPROVED** [review]

**Comment by tjruwase** [comment]
@leiwen83, apologies for the delayed review. Can you please follow the guide below to address the formatting issues?
https://github.com/microsoft/DeepSpeed/blob/master/CONTRIBUTING.md#contributing

**Comment by leiwen83** [comment]
> @leiwen83, apologies for the delayed review. Can you please follow the guide below to address the formatting issues? https://github.com/microsoft/DeepSpeed/blob/master/CONTRIBUTING.md#contributing

formatting issue is fixed in my latest push.

**Comment by leiwen83** [comment]
@tjruwase Is there something else need to be fixed for this patch? Or shall we merge this PR?

Thx

**Comment by tjruwase** [comment]
@leiwen83, apologies, I forgot about this PR. I will add to merge queue once CI completes.

**Comment by leiwen83** [comment]
> @leiwen83, apologies, I forgot about this PR. I will add to merge queue once CI completes.

Hi @tjruwase ,

Seem CI is finished, shall we merge patch?

Thx

**Comment by tjruwase** [comment]
@leiwen83, I have scheduled for auto-merge.

**Comment by leiwen83** [comment]
> @leiwen83, there are actually CI failures which I did not notice earlier. image
> 
> The stage 1/2 [load_state](https://github.com/microsoft/DeepSpeed/blob/94c7233a8bb51e068ff8dd5d3e03f2e9b5ab248e/deepspeed/runtime/zero/stage_1_and_2.py#L2143) function needs to be updated. Thanks!

@tjruwase This bug is fixed in the latest push.

**Comment by tjruwase** [comment]
@leiwen83, still CI failure due to missing here:
https://github.com/microsoft/DeepSpeed/blob/94c7233a8bb51e068ff8dd5d3e03f2e9b5ab248e/deepspeed/runtime/zero/mics.py#L464

**Comment by leiwen83** [comment]
> @leiwen83, still CI failure due to missing here:
> 
> https://github.com/microsoft/DeepSpeed/blob/94c7233a8bb51e068ff8dd5d3e03f2e9b5ab248e/deepspeed/runtime/zero/mics.py#L464

@tjruwase This bug is fixed. I test with tests/unit/checkpoint/test_mics_optimizer.py::TestMiCSCheckpoint, and it pass in my local.

**Comment by leiwen83** [comment]
@tjruwase 
I see automerge is blocked... Is it related with my patch, since I see there is nccl error...
```
FAILED unit/inference/test_inference.py::TestMPSize::test[fp32-gpt-neo] - ValueError: `do_sample` is set to `False`. However, temperature is set to 0.9 -- this flag is only used in sample-based generation modes. Set `do_sample=True` or unset temperature to continue.
FAILED unit/inference/test_inference.py::TestMPSize::test[fp16-gpt-neo] - ValueError: `do_sample` is set to `False`. However, temperature is set to 0.9 -- this flag is only used in sample-based generation modes. Set `do_sample=True` or unset temperature to continue.
FAILED unit/hybrid_engine/test_he_llama.py::TestHybridEngineLlama::test_functionality[huggyllama/llama-7b-bsz=1] - RuntimeError: Workspace can't be allocated, not enough memory
FAILED unit/hybrid_engine/test_he_llama.py::TestHybridEngineLlama::test_functionality[huggyllama/llama-7b-bsz=2] - RuntimeError: NCCL error in: ../torch/csrc/distributed/c10d/NCCLUtils.hpp:121, unhandled cuda error, NCCL version 2.14.3
ncclUnhandledCudaError: Call to CUDA function failed.
Last error:
Cuda failure 'out of memory'

```

**Comment by tjruwase** [comment]
@leiwen83, no, I don't think this is caused by your PR. It seems to be unrelated, and we are investigating it. Apologies for the delay.
